### PR TITLE
plan: add hint to force to choose hash join.

### PR DESCRIPTION
**Pull request by winoros**
for #5282 
PTAL @hanfei1991 @coocood @lamxTyler

### CODE REVIEW

File: parser/parser.y [hunk 1]
```diff
@@ -421,6 +421,7 @@ import (
 	statsHistograms "STATS_HISTOGRAMS"
 	statsBuckets    "STATS_BUCKETS"
 	tidb		"TIDB"
+	tidbHASHJOIN    "TIDB_HASHJOIN"
```
Comment (hanfei1991) [inline parser/parser.y]: As convention, we should use TIDB_HJ

Comment (shenli) [reply parser/parser.y]: s/tidbHASHJOIN/tidbHashJoin/

File: plan/join_reorder.go [hunk 2]
```diff
@@ -29,7 +30,7 @@ func tryToGetJoinGroup(j *LogicalJoin) ([]LogicalPlan, bool) {
 	// 2. not inner join
 	// 3. forced merge join
 	// 4. forced index nested loop join
-	if j.reordered || !j.cartesianJoin || j.preferMergeJoin || j.preferINLJ > 0 {
+	if j.reordered || !j.cartesianJoin || bits.OnesCount(j.preferJoinType) > 0 {
```
Comment (lamxTyler) [inline plan/join_reorder.go]: Should we update the comment?

File: executor/join_test.go [hunk 3]
```diff
@@ -147,6 +147,8 @@ func (s *testSuite) TestJoin(c *C) {
 
 	// Test that two conflict hints will return error.
 	_, err = tk.Exec("select /*+ TIDB_INLJ(t) TIDB_SMJ(t) */ * from t join t1 on t.a=t1.a")
+	_, err = tk.Exec("select /*+ TIDB_INLJ(t) TIDB_HASHJOIN(t) */ from t join t1 on t.a=t1.a")
+	_, err = tk.Exec("select /*+ TIDB_SMJ(t) TIDB_HASHJOIN(t) */ from t join t1 on t.a=t1.a")
```
Comment (lamxTyler) [inline executor/join_test.go]: The `err` will be overwritten, so check just once is not enough.

File: plan/logical_plan_builder.go [hunk 4]
```diff
@@ -1800,10 +1812,20 @@ func (b *planBuilder) buildSemiJoin(outerPlan, innerPlan LogicalPlan, onConditio
...
+			joinPlan.preferJoinType |= preferHashJoin
+		}
 		// semi join's outer is always the left side.
 		if b.TableHints().ifPreferINLJ(outerAlias) {
-			joinPlan.preferINLJ = preferLeftAsOuter
+			joinPlan.preferJoinType = preferLeftAsIndexOuter
+		}
+		// If there're multiple join hints, they're conflict.
+		if bits.OnesCount(joinPlan.preferJoinType) > 1 {
+			b.err = errors.New("Optimizer Hints is conflict")
```
Comment (zz-jason) [inline plan/logical_plan_builder.go]: make error message more specific ? like: Optimizer Hints conflict, at most one of {TIDB_INLJ, TIDB_SMJ, TIDB_HASHJOIN} join hints should be specified.

File: plan/logical_plan_builder.go [hunk 5]
```diff
@@ -1800,10 +1812,20 @@ func (b *planBuilder) buildSemiJoin(outerPlan, innerPlan LogicalPlan, onConditio
 	if b.TableHints() != nil {
```
Comment (XuHuaiyu) [inline plan/logical_plan_builder.go]: can this branch be extracted as a function.

Comment (XuHuaiyu) [reply plan/logical_plan_builder.go]: @winoros this branch can be extracted as a function?

Comment (winoros) [reply plan/logical_plan_builder.go]: i think it's not too much need to do this.

**Comment by XuHuaiyu** [comment]
LGTM

**Comment by coocood** [comment]
LGTM

**Comment by shenli** [comment]
@winoros Remind to update the document.

**Pull request merged**